### ISSUE TITLE
Make it possible to push local block styles to global block styles

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#44361]
When editing a block in the post or site editor, it would be nice if there was a way to push the styles for that block to global styles so that all blocks across the site are updated. It could work like this:

The button in "Advanced" adds the "unsaved changes" dot:

<img width="1876" alt="Button in Advanced" src="https://user-images.githubusercontent.com/1204802/200558155-ecfc31c6-0207-44cb-bff7-d4c17d3eae54.png">

That means publishing now surfaces an option to "publish" the styles that were pushed:

<img width="1876" alt="Multi entity saving" src="https://user-images.githubusercontent.com/1204802/200558264-8658cd3c-83ed-46f7-ba89-1a79a927c0ac.png">

I still think the nested menus option would be the clearest path forward, which would manage prominence. Using the option here would also add the "unsaved dot" and require multi entity saving:

<img width="905" alt="Nested menus" src="https://user-images.githubusercontent.com/1204802/200558413-03f69d53-2817-4346-965c-4aca7cb29708.png">

A smaller option would be to accept a longer ellipsis menu:

<img width="905" alt="Just menus" src="https://user-images.githubusercontent.com/1204802/200558452-1a3c28d6-be87-4d48-9c1d-fa58aff9555d.png">

As a summary:

* Pushing local changes to global queues it for multi entity saving
* A minimal first version could put these options in the block toolbar ellipsis menu, ideally the nested version if possible
* An inspector affordance in the Advanced section could also work
* See also comments for ideas on [per-panel style pushing](https://github.com/WordPress/gutenberg/issues/44361#issuecomment-1307409623) to be explored separately

---

Issue updated Nov 9. Props @javierarce for mockups.

**Comment by tanjoymor** [comment]
Love this idea!

**Comment by javierarce** [comment]
Here are some proposals for this feature:

First of all, we could include a new item in the contextual ellipsis menu. It could be a single item (left), or we could combine it into a submenu (right) thanks to the addition of the paste style action from [this issue](https://github.com/WordPress/gutenberg/issues/44418).

![image](https://user-images.githubusercontent.com/4933/197499726-1dbbb7a3-8a49-4836-99f7-851aecc221f8.png)

Another idea could be placing the push functionality in the block inspector:

We could have the button in the Advanced section. After pushing the styles a snackbar message would inform users that the Global Styles have been updated. A link to review the changes could be offered (but probably only in the Site Editor)

![image](https://user-images.githubusercontent.com/4933/197495296-000b6392-95b5-44f9-a03b-68dc6bd43b7c.png)

Iterating on that last idea, we could offer a summary section for the styles. It would allow users to inspect the styles for the block, copy them, or make them global. I remember seeing a ticket about this inspector summary, but I couldn't find it.

![image](https://user-images.githubusercontent.com/4933/197498016-d7a0093e-d9fe-45fc-beb2-76160cc08982.png)

Something important to consider: after the styles are pushed there should be possible to undo the change with `ctrl+z`

[Figma file](https://www.figma.com/file/EWsMbr0GXrMWpxHaBxpYBc/Push-Local-Block-Styles?node-id=1%3A2)

**Comment by jameskoster** [comment]
@javierarce do you think the change proposed in https://github.com/WordPress/gutenberg/issues/40204 would affect the Inspector design?

**Comment by javierarce** [comment]
Thanks for pointing that PR out, I had forgotten about it. 

I think the styles section (with or without the style summary) could appear at the bottom of the Appearance tab under a "Styles" label, so in essence, it would be similar to the design above.

Another idea I had testing that PR could be adding an inspector tab with the styles summary, but I'm not sure if placing the push action there could be confusing since everything related to managing styles is done through the Appearance tab.

**Comment by jameskoster** [comment]
For the menu, I quite like the context that the dedicated Styles flyout provides. Do you think the "Push to Global Styles" action needs a description, or a confirmation dialog that appears on click explaining what that action actually means? I'm also wondering if we should factor the inverse action (revert to default styles) into this design?

I'm torn on the style summary, it feels a bit duplicative to me. When do you imagine this would be most useful?

**Comment by jasmussen** [comment]
> Something important to consider: after the styles are pushed there should be possible to undo the change with ctrl+z

As an alternative, we could make "dirty" global styles, so that the Publish button changes to "Save" and invokes the multi entity saving interface. That would act as a confirm, mitigating the need for undo, as I imagine that might be tricky, as undo normally doesn't do any saving at all, wheres this would be both push and save for a separate entity.

This multi entity saving panel might also be where any changes made can be summarized. For [theme switching](https://make.wordpress.org/design/2021/10/08/adventures-in-block-theme-switching/), we want to summarize every change made before applied, so that you can check/uncheck the changes you want. For example, switching from TT2 to TT3, you might want to keep the patterns and colors from TT2 but switch to the new base layout from TT3, then you could uncheck colors and patterns in the theme switching dialog (currently multi entity saving sidebar, but could become a modal). 

This same interface could be used for this copying. That is to say, we just don't omit this summary part for now, and rely on the fact that in the future, the multi entity saving interface will show checkboxes and a summary. Makes sense?

Regarding the dropdown flyout, I would echo Jay, love the context it brings, and the fact that it lets us avoid a new panel. My only concern there is that dropdown flyouts are tricky to implement. @ntsekouras if you have a moment, how hard would such a flyout be to implement?

<img width="441" alt="Screenshot 2022-10-25 at 12 08 11" src="https://user-images.githubusercontent.com/1204802/197746095-86e7b754-1cfd-4fc0-85c9-6676609010da.png">

If very hard, then perhaps we create a section of the main ellipsis menu and have a good label + icon for it, for the near term.

**Comment by ntsekouras** [comment]
>Regarding the dropdown flyout, I would echo Jay, love the context it brings, and the fact that it lets us avoid a new panel. My only concern there is that dropdown flyouts are tricky to implement. @ntsekouras if you have a moment, how hard would such a flyout be to implement?

I don't think it would be hard to do that.. I tried something real quick and it seems it will be fine. --cc @ciampo as he's a Popover expert right now 😄 .

**Comment by jasmussen** [comment]
I like the sound of that! Let's definitely look into flyouts then. Logistical question for @javierarce then, should the flyout (Copy, paste, push) be disconnected from the main menu with 12px, just like the main dropdown is disconnected from the block toolbar?

**Comment by ciampo** [comment]
> > Regarding the dropdown flyout, I would echo Jay, love the context it brings, and the fact that it lets us avoid a new panel. My only concern there is that dropdown flyouts are tricky to implement. @ntsekouras if you have a moment, how hard would such a flyout be to implement?
> 
> I don't think it would be hard to do that.. I tried something real quick and it seems it will be fine. --cc @ciampo as he's a Popover expert right now 😄 .

That looks like a nested dropdown menu, which to my knowledge we don't currently support (or at least I think! I don't remember seeing a nested dropdown menu in gutenberg before).

The proper way to implement that would be extend the functionality of `DropdownMenu` — I'm not sure yet how complex it would be, but feel free to kickstart the conversation in a separate issue and ping us folks from components squad! I guess the hardest part will be to nail accessibility and keyboard interaction, which we'd need to explore together.

**Comment by jasmussen** [comment]
@javierarce This design feels solid:

<img width="859" alt="Screenshot 2022-11-01 at 13 55 09" src="https://user-images.githubusercontent.com/1204802/199236963-ee41f043-4c97-4e9c-815f-fa7071d3ff6e.png">

And since the devs aren't blanket dismissing a nested dropdown like this, I think this is the design we should settle on as a starting point. If we combine that with the _action of pushing a local style to global_ adding the unsaved dot, and invoking multi entity saving, it may be enough to graduate this one to dev territory. (I'd also love to see the nested submenu being detached from the parent dropdown.) What do you think? 

If you can point to the Figma file, I'd be happy to make the changes, and update the issue — I know you have a lot on your plate. Just let me know!

**Comment by ciampo** [comment]
> since the devs aren't blanket dismissing a nested dropdown like this, I think this is the design we should settle on as a starting point

While we didn't "blanket dismiss it" above, I just wanted to reiterate the fact that nested dropdown menus can be complex to implement correctly, especially in terms of accessibility, pointer and keyboard interaction. It would probably wise to also think about an alternative UI that doesn't employ nested dropdown menus, in case anything comes up.

Another couple of things to keep in mind that may be important to consider:
- we already have popovers opening inside popovers (e.g. color pickers inside gradient pickers, dropdown menus opening from floating toolbars..). Nested dropdown menus would add one (or more) additional layers of nested popovers to our UIs, making it ever more complex / less discoverable
- how would the nested dropdown translate to mobile UIs?

**Comment by colorful-tones** [comment]
Hiding the 'Push to Global Styles' in a flyout is likely going to be frustrating. Since most users will be modifying styles in the sidebar then changing the context to the flyout seems to break the flow, and it is likely users will miss the flyout (will take an awareness campaign).

I think there was some nice stuff in @javierarce design and it seems to coalesce with what is happening in #40204

![](https://user-images.githubusercontent.com/4933/197498016-d7a0093e-d9fe-45fc-beb2-76160cc08982.png)

**Comment by javierarce** [comment]
> While we didn't "blanket dismiss it" above, I just wanted to reiterate the fact that nested dropdown menus can be complex to implement correctly, especially in terms of accessibility, pointer and keyboard interaction. It would probably wise to also think about an alternative UI that doesn't employ nested dropdown menus, in case anything comes up.

The other alternative UI that I can think of is having an accordion inside the menu:

https://user-images.githubusercontent.com/4933/199273216-67727870-9ca6-4125-be9e-97755139f726.mov

It comes with their own challenges (for example, is the chevron easy to spot? is it ok to that the "Styles" item reveals the submenu with a click instead of a hover?), but I think it would solve the mobile the nesting issue @ciampo mentions.

> If you can point to the Figma file, I'd be happy to make the changes, and update the issue — I know you have a lot on your plate. Just let me know!

Thanks, @jasmussen! I can do it, but in any case [this design file](https://www.figma.com/file/EWsMbr0GXrMWpxHaBxpYBc/Push-Local-Block-Styles?node-id=1%3A2) contains the designs for this ticket including the menus. There's also [this one](https://www.figma.com/file/WNTUdJBimu1tqHJBRlYAfm/Copy-and-paste-block-styles) for the "Copy and paste block styles" improvement that is related.

**Comment by jasmussen** [comment]
> While we didn't "blanket dismiss it" above, I just wanted to reiterate the fact that nested dropdown menus can be complex to implement correctly, especially in terms of accessibility, pointer and keyboard interaction. It would probably wise to also think about an alternative UI that doesn't employ nested dropdown menus, in case anything comes up.

Sorry, didn't mean to put you on the spot. Of course we'll be able to work within what's feasible to implement and find alternatives if it balloons. But nested menus is something we can use for transforms and many other places, and a good way to manage the complexity of a growing ellipsis menu, so it's a useful tool to have in the toolbox, beyond just this.

An alternative could be to have a "Styles" label on the left, inactive, and then 3 icon buttons on the right for copy/paste/push, a bit like how the Chrome menu does Zoom:

<img width="399" alt="Screenshot 2022-11-01 at 18 43 24" src="https://user-images.githubusercontent.com/1204802/199301634-6f68b0af-5d4a-44e7-a08c-b3532259d82c.png">

Something to consider. 

One thing I wanted to elaborate on was this piece:

> As an alternative, we could make "dirty" global styles, so that the Publish button changes to "Save" and invokes the multi entity saving interface. That would act as a confirm, mitigating the need for undo, as I imagine that might be tricky, as undo normally doesn't do any saving at all, wheres this would be both push and save for a separate entity.

The thing is, we already have a way to save change to multiple entities beyond what's loaded in the current post editor. If you change a site title inside the site title block, a dot appears in the Publish box:
<img width="331" alt="Screenshot 2022-11-01 at 18 45 18" src="https://user-images.githubusercontent.com/1204802/199302059-0c82c8ee-cd91-4916-b20a-be18475e0664.png">

Clicking it invokes the multi entity saving panel:

<img width="311" alt="Screenshot 2022-11-01 at 18 45 49" src="https://user-images.githubusercontent.com/1204802/199302158-d8138dbe-62e5-4945-af9e-641c9ee746e2.png">

It seems sensible to reuse this interface for any local changes you are pushing to become global, that way we already have a confirm feature built in. Additionally, any of the "visual diffing" enhancements explored in this PR would be useful for _all_ entities being saved, not just styles, so we can consider those to be editor-wide enhancements, rather than effort spent only on this one feature. What do you think?

**Comment by jasmussen** [comment]
> Hiding the 'Push to Global Styles' in a flyout is likely going to be frustrating. Since most users will be modifying styles in the sidebar then changing the context to the flyout seems to break the flow, and it is likely users will miss the flyout (will take an awareness campaign).

I hear that, and I think there's sense in exploring an inspector affordance for it which considers the incoming tabbed interface.

However it is also an advanced feature. In Figma, for example, I use it all the time, but the feature is incredibly buried under an obtuse Edit > Copy properties menu item, or by clicking a very tiny hit area in the fill-layers and pressing ⌘C, then ⌘V on another object. 

I like to think that it's important to find the right balance here as well, one where it's ergonomic and easy for a theme developer to routinely use the feature, but still managing its prominence.

If the flyout submenu had shortcut keys visualized in the menu, would that help thread the needle?

**Comment by colorful-tones** [comment]
> However it is also an advanced feature.

I gamble it'll go from an advanced feature to a ubiquitous tool for everyday builders to leverage as part of their site design/building, but merely speculating 🔮 . For now, I'm fine with it hidden away, but would not be surprised if it is necessary to revisit the user interface further down the road 👍

**Comment by jasmussen** [comment]
@javierarce I took the liberty of [building on your work](https://github.com/WordPress/gutenberg/issues/44361#issuecomment-1288754703) to [illustrate the idea](https://www.figma.com/file/EWsMbr0GXrMWpxHaBxpYBc/Push-Local-Block-Styles?node-id=910%3A6364) I had for "marking the document as dirty" and using multi entity saving.

The concepts are essentially the same as yours, with regards to the button in Advanced, and options in the flyouts, but with a few small tweaks, then the inclusion of the multi entity saving concept. 

The button in "Advanced" adds the "unsaved changes" dot:

<img width="1876" alt="Button in Advanced" src="https://user-images.githubusercontent.com/1204802/200558155-ecfc31c6-0207-44cb-bff7-d4c17d3eae54.png">

That means publishing now surfaces an option to "publish" the styles that were pushed:

<img width="1876" alt="Multi entity saving" src="https://user-images.githubusercontent.com/1204802/200558264-8658cd3c-83ed-46f7-ba89-1a79a927c0ac.png">

I still think the nested menus option would be the clearest path forward, which would manage prominence. Using the option here would also add the "unsaved dot" and require multi entity saving:

<img width="905" alt="Nested menus" src="https://user-images.githubusercontent.com/1204802/200558413-03f69d53-2817-4346-965c-4aca7cb29708.png">

A smaller option would be to accept a longer ellipsis menu:

<img width="905" alt="Just menus" src="https://user-images.githubusercontent.com/1204802/200558452-1a3c28d6-be87-4d48-9c1d-fa58aff9555d.png">

Let me know what you think!

**Comment by colorful-tones** [comment]
@jasmussen @javierarce this is some really nice work. I'm a fan of the Advanced sidebar integration and love the "dirty" indicator flow.

I just had a small suggestion and please excuse the 5-minute design modification in Preview app, but thought it might be good to try and reinforce which styles items are dirty in the sidebar once they're modified and used @jasmussen latest image and modified it quickly. I'm not entirely happy with the circle colors/treatment and was mostly just going for a quick visual to demonstrate my idea. 

<img width="1876" alt="200558155-ecfc31c6-0207-44cb-bff7-d4c17d3eae54" src="https://user-images.githubusercontent.com/405912/200607736-93e47a6c-b10e-4529-a51d-c689f1c8fc3e.png">

**Comment by jameskoster** [comment]
I think that's a nice idea @colorful-tones, and could facilitate per-panel-pushing and resetting. The latter will be particularly important if you've extensively modified a block and want to push everything _except_ your border changes (as an example).

**Comment by jasmussen** [comment]
I'll remember to share the Figma links more often — thanks so much for the added exploration!

Conceptually I like the idea of per-panel style pushing, but I'm a bit concerned with unread-dot fatigue. When I see such a dot, I want to address the issue it's pointing me to — which is why it's good for the multi entity saving interface, and personally frustrating for an interface where I can't make it go away, such as in the MacOS preferences:

<img width="309" alt="Screenshot 2022-11-09 at 09 38 37" src="https://user-images.githubusercontent.com/1204802/200781189-6c808f19-f4c8-4b03-bc6d-ec51e2b47215.png">

_I'm not using iCloud._

In this case, I might be confused at dots appearing for local changes I made. However, and to expand on Jay's point above, there might be an opportunity to explore embedding this feature inside the panel ellipsis menu, this one:

<img width="369" alt="Screenshot 2022-11-09 at 09 41 19" src="https://user-images.githubusercontent.com/1204802/200781651-ce55f276-8aa4-4f72-9af5-dd94a4804126.png">

Not sure how it would look, but there would potentially be room for a section similar to what we put in the Advanced section for now. What do you think?

**Comment by jameskoster** [comment]
I think it makes a lot of sense and would make this feature much more flexible. The menu item can probably be the same as the button you have in the Advanced panel: "Push changes to global styles". We could include the panel name "Push dimensions changes to global styles", but it's probably not necessary.

That said, I do think it's fine to have a whole-panel push affordance to begin with, and add the per-panel pushing as a follow up. Like you said, the dot isn't perfect and there are other options we can explore to indicate when a panel has departed from global styles.

**Comment by jasmussen** [comment]
Cool, I'll update this issue with the newest mockups, but point to the comments that discuss potential followups.

**Comment by jasmussen** [comment]
This issue has been updated with mockups, and is ready for a dev to pick it up!

**Comment by jasmussen** [comment]
For any dev picking this up, be sure to also see #44418.

**Comment by ramonjd** [comment]
I'm having a look at this for now until it gets too hard or I run out of time 😄

**Comment by kevin940726** [comment]
I think the difficult part of this is the same as in #44418: what attributes are considered "styles" to copy/push. There are some discussions in my draft PR https://github.com/WordPress/gutenberg/pull/45477 too.

**Comment by ramonjd** [comment]
> I think the difficult part of this is the same as in https://github.com/WordPress/gutenberg/issues/44418: what attributes are considered "styles" to copy/push. There are some discussions in my draft PR https://github.com/WordPress/gutenberg/pull/45477 too.

Great point. I was just thinking about this and I believe the caveat must be that only styles available to global styles are valid.

For example, see the Cover block global styles here:

<img width="287" alt="Screen Shot 2022-11-23 at 12 59 02 pm" src="https://user-images.githubusercontent.com/6458278/203455943-151fa594-8690-488a-9690-65a84fe9f9fd.png">

It does not include the custom controls in the Cover block for min height and overlay background, since these are block attributes not styles that can be defined in theme.json.

Also, we need to think about how to deal with presets, which are usually stored in the block's top-level attributes, and not in the style object. E.g., 

```json
{
    "fontSize": "a-font-slug",
   "styles": {
       "typography": { ... },
       ...
   }
}
```

It might be easier to communicate this constraint in the sidebar for this issue, but trickier for copy/paste functionality.

**Edit**: for copy/paste maybe we could limit it to block supports styles, e.g., typography, dimensions, color... 🤷

**Comment by ramonjd** [comment]
I've been looking into this (rather slowly as I've had to dive down a few unfamiliar rabbit holes), and am just jotting down some notes. 

I think the first version should target the site editor only. 

This is because a `GlobalStylesContext` already exists (to get and set current, merged global styles) and we can see the style updating dynamically. In the post editor, the functionality to update global styles in the DOM isn't available yet, and could therefore be a follow up.

Furthermore, in the context of the site editor, there might be less confusion about which styles are being pushed. Many blocks have custom "style" attributes, e.g., overly color for the Cover block, column count for the Columns block, which are not compatible with theme.json/global styles. 

Also, and probably an esoteric point, we'll want to make sure that we don't overwrite any other style properties when updating an individual property. For example, if we change the top border color for a Group block, we want to make sure we're only pushing that one value globally, and nothing else.

Another thought I had was about block attributes themselves. If a block's styles are pushed globally, should that block retain the styles at the block level? My instinct is to say "yes" to avoid unforeseen styling issues if global styles are cleared. Just a thought I had.

Down the line, we might want to leave the door open to allowing individual style properties to be pushed. So a user might only want their changes to `typography.letterSpacing` to be pushed globally, and not their changes to color or whatever. 

#45961 is a very crude prototype, and I'm not sure I'll have enough time to get something acceptable working for the purposes of [Phase 2](https://github.com/WordPress/gutenberg/issues/33094) in the next couple of weeks. 

If it comes down to that I'll unassign myself and add any learnings here.

**P.S.** There might be a few things we can learn from https://github.com/WordPress/gutenberg/pull/34178, or maybe not. The aim of that PR was to allow blocks to access global styles. Just noting here so I don't forget 😄

**Comment by annezazu** [comment]
In a recent hallway hangout with the FSE Outreach Program, we discussed this issue alongside https://github.com/WordPress/gutenberg/issues/44412. How will these two interconnect, meaning if I set up custom CSS and try to push globally, how will that be handled?

**Comment by ramonjd** [comment]
> How will these two interconnect, meaning if I set up custom CSS and try to push globally, how will that be handled?

Thanks for the question; it's a good one. 

I believe, based only on what I've learned so far in tinkering with https://github.com/WordPress/gutenberg/pull/45961, that the mechanics will be slightly different. 

To push custom CSS globally, I suspect we'd have to parse the custom CSS for the limited style properties and sub properties that global styles supports - typography, color, border and so on.  cc @glendaviesnz for fact check

Saving them to the user styles entity via the global style context will be the same I'm assuming.

**Comment by ramonjd** [comment]
After discussing with @noisysocks I've closed https://github.com/WordPress/gutenberg/pull/45961#issuecomment-1343768343 with comments.

A follow up PR will be available at some point soon - one that will be, I suspect, less complex 😄 

A few further things we'll need to keep in mind that came out of the discussion:

- There's still the question of what needs to be changed. Newly-changed block attributes or all of them at once? What if a block has had its local styles changes on a previous date? My instinct is that it'd be okay to push all local block's eligible attribute styles and preset values at one regardless of when they were added.
- When pushing to global styles, the local block styles can be nullified
- Examine the undo levels: there should be one undo for a push operation (?)
- How do we deal with fallbacks? E.g., if a block's local styles are reset and a user pushes the styles to global styles, are the user styles reset to default (theme.json) or any previously-entered value in the global styles panel (might be hard)

**Comment by jasmussen** [comment]
> There's still the question of what needs to be changed. Newly-changed block attributes or all of them at once? What if a block has had its local styles changes on a previous date? My instinct is that it'd be okay to push all local block's eligible attribute styles and preset values at one regardless of when they were added.

I could be missing something, but what would be the use case of _not_ pushing all styles?

> When pushing to global styles, the local block styles can be nullified

Can you elaborate? Is this because you are _moving_ the local styles to being global? If so, that seems valid enough only after you actually save those custom styles.

> Examine the undo levels: there should be one undo for a push operation (?)

Yes. Question, would this undo persist if you save the global styles? And in that case, would the undo also undo on the global level?

> How do we deal with fallbacks? E.g., if a block's local styles are reset and a user pushes the styles to global styles, are the user styles reset to default (theme.json) or any previously-entered value in the global styles panel (might be hard)

I'm having a hard time parsing the example. If a block's local styles are reset, what would they then source from? And wouldn't those styles sourced, if pushed, then be the new standard, saved in the user style?

**Comment by ramonjd** [comment]
Sorry, end of day brain dump without rereading 😄 

> I could be missing something, but what would be the use case of not pushing all styles?

The only case I could come up with was when a user wants some local block styles, but not all of them, pushed. For example, I've made some changes to a local block, and I then return tomorrow and make some more. I only want the second round of changes I make in real-time to apply globally.

It's a contrived example, and I do think it's totally okay to push them all at once, but I wanted to raise it to indicate that we had at least considered it.

Here's an alternative PR that pushes all styles: https://github.com/WordPress/gutenberg/pull/46446 (kudos to @noisysocks for working on that)

> When pushing to global styles, the local block styles can be nullified

The assumption is that local block styles that have been pushed globally no longer need to be applied to the local block. 

So after the local styles are pushed, we can delete them from the local block styles, since they'll be inherited from global styles.

There is a downside to this: if a user resets global styles, expecting that the local block styles will persist. If we remove them when pushing to global styles, they'll be gone forever. cc @noisysocks 

>  Yes. Question, would this undo persist if you save the global styles? And in that case, would the undo also undo on the global level?

I think so, yes. We're currently looking into it. @noisysocks started investigating yesterday, but the work of undo is voodoo so we might have to come back to you once we've verified how it will and how it can work.

> I'm having a hard time parsing the example. If a block's local styles are reset, what would they then source from? And wouldn't those styles sourced, if pushed, then be the new standard, saved in the user style?

Sorry, it is a bit strangely worded. 

I tied myself in a knot over this one, and I believe it's not a valid concern. So you can ignore me. 

To explain, I was pondering over the following scenario:

1. A user applies some global styles to a paragraph block via the global styles panel. 
2. A user then edits a paragraph block and pushes its local block styles globally via the block inspector, overwriting the former.
3. A user returns to that local paragraph block, and resets all the styles, then pushes globally.

So I was asking myself what "reset" meant here and whether any user global styles also had to be "reset" back to what they were. 

But now I think about it, the point seems invalid: if I reset a local block's styles and then push globally, I'd expect that all blocks of that type will look the same as the current block, that is, with reset styles.
